# Metabarcoding analysis of 16S region of BSM24 sequences using Obitools 4.4.0 

Set the directory and Obitools version.

        cd /home/scc/cramos/16S

        module load obitools/4.4.0        

## 1. FASTQC analysis**


FastQC (Andrews S., 2010) purpose is to offer an easy-to-use tool for performing basic quality checks on raw sequencing data generated by high-throughput sequencing workflows. It includes a set of modular analyses designed to quickly highlight any potential issues in the data that should be addressed before proceeding with downstream analysis.

        module load fastqc
        
        gunzip -k 16S_sequences_foward.fastq.gz gunzip -k 16S_sequences_reverse.fastq.gz
        
        fastqc 16S_sequences_foward.fastq.gz 16S_sequences_reverse.fastq.gz

## 2. Obitools analysis

### 2.1 Obipairing

        obipairing --min-identity=0.9 --min-overlap=20 -F 16S_sequences_forward.fastq -R 16S_sequences_reverse.fastq  > results/consensus_2.fastq

min-overlap: minimum overlap between both the reads to consider the alignment (default: 20)
min-identity: minimum identity between overlapped regions of the reads to consider the alignment (default: 0.900000)

        obicount results/consensus_2.fastq

Sequences number: **57.976.189**

### 2.2 Obigrep

        obigrep -p 'annotations.mode != "join" results/consensus_2.fastq > results/assembled.fastq

-p 'annotations.mode != "join" means if the value of the mode annotation of a sequence is different from join, then the corresponding sequence record should be kept in the output.

Sequences number: **54.161.789**

### 2.3 Obimultiplex

        obimultiplex -s 16S_new.txt -u results/unidentified_new.fastq results/assembled.fastq > results/assembled_assigned.fastq
      
Allowed-mismatches: Used to specify the number of errors allowed for matching primers. (default: -1)

Sequences number: **46.377.578**

### 2.4 Obiuniq to merge identical sequencies 

        obiuniq -m sample results/assembled_assigned.fastq > results/assembled_assigned_uniq.fasta

merge | -m: Adds a merged attribute containing the list of sequence record ids merged within this group.

Sequences number: **652.984**

### 2.5 Obigrep to deleted ambiguous bases

Before comparing with the reference database, it is commomn to find sequences with ambiguous bases represented with IUPAC codes as N, M, K, R, Y, S, W, B, D, H or V.
The objetive of this part of the code is to delete all the bases but A, C, G, T. 


        obigrep -p 'sequence =~ "^[ACTGactg]+$"' results/assembled_assigned_uniq.fasta > results/clean_sequences.fasta

Sequences number: **322.023**

### 2.6 Obiannotate

        obiannotate -k count -k merged_sample results/clean_sequences.fasta > results/clean_annotated.fasta

### 2.7 Obiclean for chimeras

        obiclean -s sample -r 0.1 --detect-chimera -H results/clean_annotated.fasta > results/cleaned_chimeras_0.1.fasta

Obiclean can be run in filter mode, allowing a sequence to be removed from the resulting sequence set if it is considered artifactual in all samples where it appears. Artifactual                sequences are those classified as internal or chimeric. This filtering is done by setting the -H option.

Sequences number: **147.513**

### 2.8 Obigrep to delete singletons 

        obigrep -p 'sequence.Count() == 1' results/cleaned_chimeras_0.1.fasta

This command extracts from a .fasta file only the sequences that appear a single time in the dataset. These single-occurrence sequences are often considered:
- noise (artifacts from PCR or sequencing)
- rare species.

        obigrep -c 10  results/cleaned_chimeras_0.1.fasta > results/no_singleton_0.1.fasta

-c <COUNT>: selects the sequence records for which the number of occurrences (i.e the count attribute) is equal to or greater than the defined minimum count.

sequences number: **10.911**

According to Walker et al, 2023, sequences that were outside the expected barcode length (< 40 or > 140 bp) were removed. we apply obigrep for this purpose.

       obigrep -l 40 -L 140 results/no_singleton_0.1.fasta > results/length_40/length_40.fasta  
        
variants= **6.364**

reads= 34.777.438

## 3. Sequences taxonomic assignment

Using the reference database (database.fasta) and the full NCBI taxonomy (ncbitaxo.tgz) we assign taxa to the sequences

        obitag -t ncbitaxo.tgz -R database/database.fasta results/length_40/length_40.fasta > results/length_40/taxo_40.fasta

        obiannotate  --delete-tag=obiclean_head --delete-tag=obiclean_headcount --delete-tag=obiclean_internalcount --delete-tag=obiclean_samplecount --delete-tag=obiclean_singletoncount results/length_40/taxo_40.fasta > results/length_40/taxo_red_40.fasta

 MOTUs (Molecular operational taxonomic units) abundances for each PCR was obtained using the merge_sample attribute and the obiclean_weight attribute. 
The merge_sample attribute was set by obiuniq andinforms about the observed number of reads for each sequence variant in the different samples. By contyarst obiclean_weight is the number of reads assigned to each sequence variant during obiclean. Obiclean_weight is the best option to represent the sequence occurrence than the merge_sample attribute.

        obimatrix --map obiclean_weight results/length_40/taxo_red_40.fasta > results/length_40/16S_occurrency.csv

To create a csv document with information about the id, count, obitag_bestid, obitag_bestmatch, taxid, sequence among others attributes, we use obicsv. The -i and -s options include the sequence identifier and the sequence itself in the output CSV file.
        
         obicsv --auto -i -s results/length_40/taxo_red_40.fasta > results/length_40/16S_MOTUS.csv

## 4. Building database. 
`cd data/scc/cramos/16S/database`

`wget -nH --cut-dirs=6 -A 'STD_*.dat.gz' -R 'STD_HUM*.dat.gz','STD_ENV*.dat.gz','STD_PLN*.dat.gz'  -m -np ftp://ftp.ebi.ac.uk/pub/databases/ena/sequence/snapshot_latest/std/`

We donÂ´t select sequences from Humans, environment and plants to make the database smaller.

`obipcr -e 2 -l 30 -L 150 --forward CGAGAAGACCCTATGGAGCT --reverse CCGAGGTCRCCCCAACC --no-order database/ > pcr_16S.fasta`

`obigrep -t database/taxdump.tar.gz -A taxid --require-rank species --require-rank genus --require-rank family pcr_16S.fasta > db_16S.fasta` 

`obiuniq -c taxid db_16S.fasta > db_uniq.fasta`

`obirefidx -t database/taxdump.tar.gz db_uniq.fasta > db_indexed.fasta`

`obitag -t ncbitaxo.tgz -R db_indexed.fasta results/length_40/taxo_red_40.fasta > results/length_40/assembled_taxo.fasta`

`obiannotate  --delete-tag=obiclean_head --delete-tag=obiclean_headcount --delete-tag=obiclean_internalcount --delete-tag=obiclean_samplecount --delete-tag=obiclean_singletoncount results/length_40/assembled_taxo.fasta > results/length_40/final_taxa_16S.fasta`

`obimatrix --map obiclean_weight results/length_40/final_taxa_16S.fasta > results/length_40/16S_occurrency.csv`

`obicsv --auto -i -s results/length_40/final_taxa_16S.fasta > results/length_40/16S_motus.csv`

# Distribution of the counts per sequences
`obicsv -k count results/length_40/taxo_red_40.fasta | tail -n +2 | sort -n | uniq -c | awk '{print $2,$1}'> results/length_40/distribution_count_per_sequences.csv`

Rstudio analysis

![Rplot](https://github.com/user-attachments/assets/900947ec-8343-4e07-805c-92620f202c98)

Mayority of the sequences are repeated 2,3,4 times...More than 13 times is weird so we show a histogram "zoom it"

![Rplot01](https://github.com/user-attachments/assets/dfb2673a-fa89-466c-b0ed-a5f3c0ae4f8d)

Calculating the mean and meadian of the distribution 








