# Metabarcoding analysis of 16S region of BSM24 sequences using Obitools 4.4.0 

Set the directory and Obitools version.

        cd /home/scc/cramos/16S

        module load obitools/4.4.0        

## 1. FASTQC analysis


FastQC (Andrews S., 2010) purpose is to offer an easy-to-use tool for performing basic quality checks on raw sequencing data generated by high-throughput sequencing workflows. It includes a set of modular analyses designed to quickly highlight any potential issues in the data that should be addressed before proceeding with downstream analysis.

        module load fastqc
        
        gunzip -k 16S_sequences_foward.fastq.gz gunzip -k 16S_sequences_reverse.fastq.gz
        
        fastqc 16S_sequences_foward.fastq 16S_sequences_reverse.fastq

Results of this analysis are stored in folder FastQC files.

## 2. Obitools 4.4.0  analysis

### 2.1 Obipairing

        obipairing --min-identity=0.8 --min-overlap=10 -F 16S_sequences_forward.fastq -R 16S_sequences_reverse.fastq  > results/consensus.fastq

Obipairing allows to discard sequences with low alignment quality. Parameters set: min-overlap (minimum overlap between both the reads to consider the alignment (default: 20)) and min-identity (minimum identity between overlapped regions of the reads to consider the alignment (default: 0.900000))

        obicount results/consensus.fastq

Sequences number: **57.976.189**

To check the first 6 sequences of the dataset:

        head -n 6 results/consensus.fastq

### 2.2 Obigrep

Unpaired reads have the attribute mode = "join" and are not reliable for downstream analysis. You can remove them using -p obigrep with this condition:

        obigrep -p 'annotations.mode != "join" results/consensus.fastq > results/assembled.fastq

-p 'annotations.mode != "join" means if the value of the mode annotation of a sequence is different from join, then the corresponding sequence record should be kept in the output.

Sequences number: **54.161.789**

### 2.3 Obimultiplex

        obimultiplex -s 16S_ngsfile.txt -u results/unidentified.fastq results/assembled.fastq > results/assembled_assigned.fastq
      
Allowed-mismatches: Used to specify the number of errors allowed for matching primers (default: -1)

Sequences number: **46.377.578**

### 2.4 Obiuniq to merge identical sequencies. Dereplication

        obiuniq -m sample results/assembled_assigned.fastq > results/assembled_assigned_uniq.fasta

merge | -m: Adds a merged attribute containing the list of sequence record ids merged within this group.

Sequences number: **652.984**

### 2.5 Obigrep to delete ambiguous sequences

        obigrep -p 'sequence =~ "^[actg]+$"' results/assembled_assigned_uniq.fasta > results/clean_sequences.fasta
        
### 2.5 Obiannotate

        obiannotate -k count -k merged_sample results/clean_sequences.fasta > results/clean_annotated.fasta

### 2.6 Obiclean for chimeras

        obiclean -s sample -r 0.1 --detect-chimera -H results/clean_annotated.fasta > results/cleaned_chimeras_0.1.fasta

Obiclean can be run in filter mode, allowing a sequence to be removed from the resulting sequence set if it is considered artifactual in all samples where it appears. Artifactual                sequences are those classified as internal or chimeric. This filtering is done by setting the -H option.

Sequences number: **147.513**

### 2.7 Obigrep to delete singletons 

        obigrep -p 'sequence.Count() == 1' results/cleaned_chimeras_0.1.fasta

This command extracts from a .fasta file only the sequences that appear a single time in the dataset. These single-occurrence sequences are often considered:
- noise (artifacts from PCR or sequencing)
- rare species.

        obigrep -c 10  results/cleaned_chimeras_0.1.fasta > results/no_singleton_0.1.fasta

-c <COUNT>: selects the sequence records for which the number of occurrences (i.e the count attribute) is equal to or greater than the defined minimum count.

sequences number: **12.840**

### 2.8 Grep by lenght

According to Walker et al, 2023, sequences that were outside the expected barcode length (< 40 or > 140 bp) were removed so we apply obigrep for this purpose.

       obigrep -l 40 -L 140 results/no_singleton_0.1.fasta > results/length_40/length_40.fasta  
        
variants= **8.096**

reads= 34.777.438

## 3. Sequences taxonomic assignment

Using the reference database (database.fasta) and the full NCBI taxonomy (ncbitaxo.tgz) we assign taxa to the sequences

        obitag -t ncbitaxo.tgz -R database/database.fasta results/length_40/length_40.fasta > results/length_40/taxo_40.fasta

        obiannotate  --delete-tag=obiclean_head --delete-tag=obiclean_headcount --delete-tag=obiclean_internalcount --delete-tag=obiclean_samplecount --delete-tag=obiclean_singletoncount results/length_40/taxo_40.fasta > results/length_40/taxo_red_40.fasta

MOTUs (Molecular operational taxonomic units) abundances for each PCR was obtained using the merge_sample attribute and the obiclean_weight attribute. 
The merge_sample attribute was set by obiuniq andinforms about the observed number of reads for each sequence variant in the different samples. By contyarst obiclean_weight is the number of reads assigned to each sequence variant during obiclean. Obiclean_weight is the best option to represent the sequence occurrence than the merge_sample attribute.

        obimatrix --map obiclean_weight results/length_40/taxo_red_40.fasta > results/length_40/16S_occurrency.csv

To create a csv document with information about the id, count, obitag_bestid, obitag_bestmatch, taxid, sequence among others attributes, we use obicsv. The -i and -s options include the sequence identifier and the sequence itself in the output CSV file.
        
         obicsv --auto -i -s results/length_40/taxo_red_40.fasta > results/length_40/16S_MOTUS.csv

## 4. Building the reference database 

        cd data/scc/cramos/16S/database

        wget -nH --cut-dirs=6 -A 'STD_*.dat.gz' -R 'STD_HUM*.dat.gz','STD_ENV*.dat.gz','STD_PLN*.dat.gz'  -m -np ftp://ftp.ebi.ac.uk/pub/databases/ena/sequence/snapshot_latest/std/

We donÂ´t select  humans, environment and plants sequences to make the database smaller.

        obipcr -e 2 -l 30 -L 150 --forward CGAGAAGACCCTATGGAGCT --reverse CCGAGGTCRCCCCAACC --no-order database/ > pcr_16S.fasta

        obigrep -t database/taxdump.tar.gz -A taxid --require-rank species --require-rank genus --require-rank family pcr_16S.fasta > db_16S.fasta

        obiuniq -c taxid db_16S.fasta > db_uniq.fasta

        obirefidx -t database/taxdump.tar.gz db_uniq.fasta > db_indexed.fasta


# 5. Distribution of the counts per sequence

        obicsv -k count results/length_40/taxo_red_40.fasta  |uniq -c  |awk '{print $2,$1}'> results/length_40/distribution_count_per_sequences.csv

# 6. Rstudio analysis


<img width="680" height="325" alt="Distribution of the counts of the sequences" src="https://github.com/user-attachments/assets/3bb4a677-be1b-4857-ac95-ad86abad5d68" />

Zoom the figure in.
<img width="680" height="325" alt="Distribution of the count of the sequences ZOOM" src="https://github.com/user-attachments/assets/8950e64c-be34-40dc-83d7-1d2f1d35b1aa" />

Calculate the mean and median of the distribution of the total of sequences obtain (6.364)

Mean = 5464.714

Median = 44





